src/Language/Haskell/LSP/Test.hs

```diff
@@ -589,3 +575,7 @@ getCodeLenses tId = do
     rsp <- request TextDocumentCodeLens (CodeLensParams tId) :: Session CodeLensResponse
     case getResponseResult rsp of
         List res -> pure res
+        
+-- | Exit the server after request its shutdown
+exitServer :: Session()
+exitServer = request_ Shutdown (Nothing :: Maybe Value) >> sendNotification Exit ExitParams
```

Comment:
Does this need to be exposed to the API? The server should automatically be exited once the withSession closure finishes

Reply:
I had to expose it cause it is used in `Language.Haskell.LSP.Test.Replay`. 
I think it should be in `Session` or in a module upstream, like the server listener handler (their clean up is closely related), and reduce the number of params of `runSessionWithHandles` but it would need move quite code (about requests/reponses) from `Test`.